File: go/nucleotide-count/nucleotide_count.go
package dna

import "errors"

// Histogram is a mapping from nucleotide to its count in given DNA.
type Histogram map[rune]int

// DNA is a list of nucleotides. Choose a suitable data type.
type DNA string

// Counts generates a histogram of valid nucleotides in the given DNA.
func (d DNA) Counts() (Histogram, error) {
	var h = Histogram{
		'A': 0,
		'C': 0,
		'G': 0,
		'T': 0,
	}
	for _, r := range d {
		if _, ok := h[r]; ok {
			h[r]++
		} else {
			return h, errors.New("invalid nucleotide")
		}
	}
	return h, nil
}

File: swift/nucleotide-count/Sources/NucleotideCount/NucleotideCount.swift
class DNA {
  var map = [ "A": 0, "C": 0, "G": 0, "T": 0 ]
  init!(strand: String) {
    for c in Array(strand) {
      if map[String(c)] == nil {
        return nil
      }
      map[String(c)]! += 1
    }
  }

  func count(_ nucleotide: String) -> Int? {
    return map[nucleotide]
  }

  func counts() -> [String: Int] {
    return self.map
  }
}